### ISSUE TITLE
twisted.names dns-spoofing vulnerability (birthday paradox)

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3347]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)| @jyknight reported|
|-|-|
|Trac ID|trac#3347|
|Type|defect|
|Created|2008-07-10 15:10:54Z|
|Branch|https://github.com/twisted/twisted/tree/dns-bday-pdox-3347|

If you run a recursive t.names server, it will do an outgoing lookup for every request you pose, until the remote server responds. This allows an attacker to exploit the "birthday paradox". By flooding the server with requests for the same hostname, many possible responses will be valid.

The fix is to not make further outgoing requests for a host when a query is already in progress.

<details><summary>Searchable metadata</summary>

```
trac-id__3347 3347
type__defect defect
reporter__jknight jknight
priority__normal normal
milestone__ 
branch__branches_dns_bday_pdox_3347 branches/dns-bday-pdox-3347
branch_author__ 
status__closed closed
resolution__fixed fixed
component__names names
keywords__ 
time__1215702654000000 1215702654000000
changetime__1216757641000000 1216757641000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [24299]) Branching to 'dns-bday-pdox-3347'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [24300]) don't issue more than one request for the same query at a time

refs [#3347](https://github.com/twisted/twisted/issues/3347)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun removed owner|
|-|-|

I'm not sure I did this right.  I changed `client.Resolver` so that it won't issue multiple requests for the same query concurrently.  I could have changed `server.DNSServerFactory` so that it won't make multiple queries for the same record set when receiving multiple requests for that record set.  At first it seemed like the responsibility for dealing with this lay with `server.DNSServerFactory` than with `client.Resolver`, then for some reason I felt like it was the other way around.  Now I don't know what I think.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_4"></a>|@jyknight set owner to @exarkun|
|-|-|

It makes sense to put there to me. That way users of t.names as a DNS client are also protected. Which seems like a good feature.

Code changes look good to me, too, so +1 merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set status to `closed`|
|-|-|

(In [24305]) Merge dns-bday-pdox-3347

Author: exarkun
Reviewer: foom
Fixes: [#3347](https://github.com/twisted/twisted/issues/3347)

Change `twisted.names.client.Resolver` so that it will not put
multiple requests for the same record set onto the network at
once.  Instead, if it is asked for a record set while it is
waiting for a response from the network which will give that
record set, it will just make sure to deliver those results
to both callers once they arrive.

This is primarily beneficial because it avoids exposing the
UDP DNS client to a "birthday paradox" attack, where it can
be easier to successfully spoof a response by causing it to
issue many duplicates of a query with different message IDs.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
